Some type failures caught by this:

```
	test_log_scope.go:61: use -show-logs to present logs inline
    --- FAIL: TestLogic/default (0.10s)
        --- FAIL: TestLogic/default/json (0.16s)
        	logic_test.go:1955:
        		testdata/logic_test/json:312: SELECT '{"a": 1}'::JSONB#>>'{a}'::STRING[]
        		expected success, but found
        		pq: binary op "'{\"a\":1}'#>>ARRAY['a']": expected return type "jsonb", got: "string"
        	logic_test.go:1976: testdata/logic_test/json:317: too many errors encountered, skipping the rest of the input
        --- FAIL: TestLogic/default/array (0.55s)
        	logic_test.go:1955:
        		testdata/logic_test/array:981: SELECT ARRAY_POSITIONS(ARRAY['A','A','B','A'], 'A'), ARRAY_POSITIONS(ARRAY['A','A','B','A'], 'C')
        		expected success, but found
        		pq: function "array_positions(ARRAY['A','A','B','A'], 'A')": expected return type "string[]", got: "int[]"
        	logic_test.go:1976: testdata/logic_test/array:986: too many errors encountered, skipping the rest of the input
```

@justinj you're familiar with both arrays and json, correct? Do you know what the right way to handle these are?

i.e. the signatures of operations mismatch the output datum type, which may cause issues.

---

Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/sql/sem/tree/eval.go, line 2170 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20524#-L-ldMrhgDfD9pY1LO12:-L-mXDFJpFNWaUNiRSHw:b-33qa7e) ([raw file](https://github.com/cockroachdb/cockroach/blob/653ddf6e4a11e62f8bf99b257189098741a0ad35/pkg/sql/sem/tree/eval.go#L2170)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

make it return via `errors.Wrapf` instead of panic
</blockquote></details>

Done. This now also returns an error instead of calling panic.

---

*[pkg/sql/sem/tree/eval.go, line 2930 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20524#-L-ld3g0GMGNs7oXJbgM:-L-mXRnROobArA-uPX8v:b-94jsne) ([raw file](https://github.com/cockroachdb/cockroach/blob/653ddf6e4a11e62f8bf99b257189098741a0ad35/pkg/sql/sem/tree/eval.go#L2930)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`Equivalent` on all `types.T` instances already handles `Any`
</blockquote></details>

So this is not true for a few `types.T`. In particular, `TTable` and `TTuple`. I'm not familiar with these types so I don't know what is appropriate here. There are cases where those are returned with an `types.Any` signature. I also had to check a test for `types.Null`, because some types are not equivalent to it.

---

*[pkg/sql/sem/tree/eval.go, line 3052 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20524#-L-h4Dq46UQnXva9xCI6:-L-mXG2kKI0sylC1mf0h:b-1tuzcq) ([raw file](https://github.com/cockroachdb/cockroach/blob/653ddf6e4a11e62f8bf99b257189098741a0ad35/pkg/sql/sem/tree/eval.go#L3052)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Yes it is a good idea.
</blockquote></details>

Done in a first commit.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20524#-:-L-m_-YBOe9D7FoOcZ0x:b-gefpda)*
<!-- Sent from Reviewable.io -->

--- a/docs/generated/sql/functions.md
+++ b/docs/generated/sql/functions.md
@@ -91,33 +91,33 @@
 </span></td></tr>
 <tr><td><code>array_position(array: oid[], elem: oid) &rarr; <a href="int.html">int</a></code></td><td><span class="funcdesc"><p>Return the index of the first occurrence of <code>elem</code> in <code>array</code>.</p>
 </span></td></tr>
-<tr><td><code>array_positions(array: <a href="bool.html">bool</a>[], elem: <a href="bool.html">bool</a>) &rarr; <a href="bool.html">bool</a>[]</code></td><td><span class="funcdesc"><p>Returns and array of indexes of all occurrences of <code>elem</code> in <code>array</code>.</p>
+<tr><td><code>array_positions(array: <a href="bool.html">bool</a>[], elem: <a href="bool.html">bool</a>) &rarr; <a href="int.html">int</a>[]</code></td><td><span class="funcdesc"><p>Returns and array of indexes of all occurrences of <code>elem</code> in <code>array</code>.</p>
 </span></td></tr>
-<tr><td><code>array_positions(array: <a href="bytes.html">bytes</a>[], elem: <a href="bytes.html">bytes</a>) &rarr; <a href="bytes.html">bytes</a>[]</code></td><td><span class="funcdesc"><p>Returns and array of indexes of all occurrences of <code>elem</code> in <code>array</code>.</p>
+<tr><td><code>array_positions(array: <a href="bytes.html">bytes</a>[], elem: <a href="bytes.html">bytes</a>) &rarr; <a href="int.html">int</a>[]</code></td><td><span class="funcdesc"><p>Returns and array of indexes of all occurrences of <code>elem</code> in <code>array</code>.</p>
 </span></td></tr>
-<tr><td><code>array_positions(array: <a href="date.html">date</a>[], elem: <a href="date.html">date</a>) &rarr; <a href="date.html">date</a>[]</code></td><td><span class="funcdesc"><p>Returns and array of indexes of all occurrences of <code>elem</code> in <code>array</code>.</p>
+<tr><td><code>array_positions(array: <a href="date.html">date</a>[], elem: <a href="date.html">date</a>) &rarr; <a href="int.html">int</a>[]</code></td><td><span class="funcdesc"><p>Returns and array of indexes of all occurrences of <code>elem</code> in <code>array</code>.</p>
 </span></td></tr>
-<tr><td><code>array_positions(array: <a href="decimal.html">decimal</a>[], elem: <a href="decimal.html">decimal</a>) &rarr; <a href="decimal.html">decimal</a>[]</code></td><td><span class="funcdesc"><p>Returns and array of indexes of all occurrences of <code>elem</code> in <code>array</code>.</p>
+<tr><td><code>array_positions(array: <a href="decimal.html">decimal</a>[], elem: <a href="decimal.html">decimal</a>) &rarr; <a href="int.html">int</a>[]</code></td><td><span class="funcdesc"><p>Returns and array of indexes of all occurrences of <code>elem</code> in <code>array</code>.</p>
 </span></td></tr>
-<tr><td><code>array_positions(array: <a href="float.html">float</a>[], elem: <a href="float.html">float</a>) &rarr; <a href="float.html">float</a>[]</code></td><td><span class="funcdesc"><p>Returns and array of indexes of all occurrences of <code>elem</code> in <code>array</code>.</p>
+<tr><td><code>array_positions(array: <a href="float.html">float</a>[], elem: <a href="float.html">float</a>) &rarr; <a href="int.html">int</a>[]</code></td><td><span class="funcdesc"><p>Returns and array of indexes of all occurrences of <code>elem</code> in <code>array</code>.</p>
 </span></td></tr>
-<tr><td><code>array_positions(array: <a href="inet.html">inet</a>[], elem: <a href="inet.html">inet</a>) &rarr; <a href="inet.html">inet</a>[]</code></td><td><span class="funcdesc"><p>Returns and array of indexes of all occurrences of <code>elem</code> in <code>array</code>.</p>
+<tr><td><code>array_positions(array: <a href="inet.html">inet</a>[], elem: <a href="inet.html">inet</a>) &rarr; <a href="int.html">int</a>[]</code></td><td><span class="funcdesc"><p>Returns and array of indexes of all occurrences of <code>elem</code> in <code>array</code>.</p>
 </span></td></tr>
 <tr><td><code>array_positions(array: <a href="int.html">int</a>[], elem: <a href="int.html">int</a>) &rarr; <a href="int.html">int</a>[]</code></td><td><span class="funcdesc"><p>Returns and array of indexes of all occurrences of <code>elem</code> in <code>array</code>.</p>
 </span></td></tr>
-<tr><td><code>array_positions(array: <a href="interval.html">interval</a>[], elem: <a href="interval.html">interval</a>) &rarr; <a href="interval.html">interval</a>[]</code></td><td><span class="funcdesc"><p>Returns and array of indexes of all occurrences of <code>elem</code> in <code>array</code>.</p>
+<tr><td><code>array_positions(array: <a href="interval.html">interval</a>[], elem: <a href="interval.html">interval</a>) &rarr; <a href="int.html">int</a>[]</code></td><td><span class="funcdesc"><p>Returns and array of indexes of all occurrences of <code>elem</code> in <code>array</code>.</p>
 </span></td></tr>
-<tr><td><code>array_positions(array: <a href="string.html">string</a>[], elem: <a href="string.html">string</a>) &rarr; <a href="string.html">string</a>[]</code></td><td><span class="funcdesc"><p>Returns and array of indexes of all occurrences of <code>elem</code> in <code>array</code>.</p>
+<tr><td><code>array_positions(array: <a href="string.html">string</a>[], elem: <a href="string.html">string</a>) &rarr; <a href="int.html">int</a>[]</code></td><td><span class="funcdesc"><p>Returns and array of indexes of all occurrences of <code>elem</code> in <code>array</code>.</p>
 </span></td></tr>
-<tr><td><code>array_positions(array: <a href="time.html">time</a>[], elem: <a href="time.html">time</a>) &rarr; <a href="time.html">time</a>[]</code></td><td><span class="funcdesc"><p>Returns and array of indexes of all occurrences of <code>elem</code> in <code>array</code>.</p>
+<tr><td><code>array_positions(array: <a href="time.html">time</a>[], elem: <a href="time.html">time</a>) &rarr; <a href="int.html">int</a>[]</code></td><td><span class="funcdesc"><p>Returns and array of indexes of all occurrences of <code>elem</code> in <code>array</code>.</p>
 </span></td></tr>
-<tr><td><code>array_positions(array: <a href="timestamp.html">timestamp</a>[], elem: <a href="timestamp.html">timestamp</a>) &rarr; <a href="timestamp.html">timestamp</a>[]</code></td><td><span class="funcdesc"><p>Returns and array of indexes of all occurrences of <code>elem</code> in <code>array</code>.</p>
+<tr><td><code>array_positions(array: <a href="timestamp.html">timestamp</a>[], elem: <a href="timestamp.html">timestamp</a>) &rarr; <a href="int.html">int</a>[]</code></td><td><span class="funcdesc"><p>Returns and array of indexes of all occurrences of <code>elem</code> in <code>array</code>.</p>
 </span></td></tr>
-<tr><td><code>array_positions(array: <a href="timestamp.html">timestamptz</a>[], elem: <a href="timestamp.html">timestamptz</a>) &rarr; <a href="timestamp.html">timestamptz</a>[]</code></td><td><span class="funcdesc"><p>Returns and array of indexes of all occurrences of <code>elem</code> in <code>array</code>.</p>
+<tr><td><code>array_positions(array: <a href="timestamp.html">timestamptz</a>[], elem: <a href="timestamp.html">timestamptz</a>) &rarr; <a href="int.html">int</a>[]</code></td><td><span class="funcdesc"><p>Returns and array of indexes of all occurrences of <code>elem</code> in <code>array</code>.</p>
 </span></td></tr>
-<tr><td><code>array_positions(array: <a href="uuid.html">uuid</a>[], elem: <a href="uuid.html">uuid</a>) &rarr; <a href="uuid.html">uuid</a>[]</code></td><td><span class="funcdesc"><p>Returns and array of indexes of all occurrences of <code>elem</code> in <code>array</code>.</p>
+<tr><td><code>array_positions(array: <a href="uuid.html">uuid</a>[], elem: <a href="uuid.html">uuid</a>) &rarr; <a href="int.html">int</a>[]</code></td><td><span class="funcdesc"><p>Returns and array of indexes of all occurrences of <code>elem</code> in <code>array</code>.</p>
 </span></td></tr>
-<tr><td><code>array_positions(array: oid[], elem: oid) &rarr; oid[]</code></td><td><span class="funcdesc"><p>Returns and array of indexes of all occurrences of <code>elem</code> in <code>array</code>.</p>
+<tr><td><code>array_positions(array: oid[], elem: oid) &rarr; <a href="int.html">int</a>[]</code></td><td><span class="funcdesc"><p>Returns and array of indexes of all occurrences of <code>elem</code> in <code>array</code>.</p>
 </span></td></tr>
 <tr><td><code>array_prepend(elem: <a href="bool.html">bool</a>, array: <a href="bool.html">bool</a>[]) &rarr; <a href="bool.html">bool</a>[]</code></td><td><span class="funcdesc"><p>Prepends <code>elem</code> to <code>array</code>, returning the result.</p>
 </span></td></tr>

--- a/docs/generated/sql/operators.md
+++ b/docs/generated/sql/operators.md
@@ -11,7 +11,7 @@
 <table><thead>
 <tr><td><code>#>></code></td><td>Return</td></tr>
 </thead><tbody>
-<tr><td>jsonb <code>#>></code> <a href="string.html">string[]</a></td><td>jsonb</td></tr>
+<tr><td>jsonb <code>#>></code> <a href="string.html">string[]</a></td><td><a href="string.html">string</a></td></tr>
 </tbody></table>
 <table><thead>
 <tr><td><code>%</code></td><td>Return</td></tr>

--- a/pkg/sql/logictest/logic_test.go
+++ b/pkg/sql/logictest/logic_test.go
@@ -37,6 +37,8 @@ import (
 	"time"
 	"unicode/utf8"
 
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+
 	"github.com/pkg/errors"
 	"golang.org/x/net/context"
 
@@ -802,6 +804,11 @@ func (t *logicTest) setup(cfg testClusterConfig) {
 				},
 				Store: &storage.StoreTestingKnobs{
 					BootstrapVersion: cfg.bootstrapVersion,
+				},
+				SQLEvalContext: &tree.EvalContextTestingKnobs{
+					AssertBinaryExprReturnTypes: true,
+					AssertUnaryExprReturnTypes:  true,
+					AssertFuncExprReturnTypes:   true,
 				},
 			},
 		},

--- a/pkg/sql/sem/builtins/builtins.go
+++ b/pkg/sql/sem/builtins/builtins.go
@@ -1987,7 +1987,7 @@ CockroachDB supports the following flags:
 	"array_positions": arrayBuiltin(func(typ types.T) tree.Builtin {
 		return tree.Builtin{
 			Types:        tree.ArgTypes{{"array", types.TArray{Typ: typ}}, {"elem", typ}},
-			ReturnType:   tree.FixedReturnType(types.TArray{Typ: typ}),
+			ReturnType:   tree.FixedReturnType(types.TArray{Typ: types.Int}),
 			Category:     categoryArray,
 			NullableArgs: true,
 			Fn: func(ctx *tree.EvalContext, args tree.Datums) (tree.Datum, error) {

--- a/pkg/sql/sem/tree/eval.go
+++ b/pkg/sql/sem/tree/eval.go
@@ -1940,7 +1940,20 @@ func (s backgroundCtxProvider) Ctx() context.Context {
 var _ CtxProvider = backgroundCtxProvider{}
 
 // EvalContextTestingKnobs contains test knobs.
-type EvalContextTestingKnobs struct {}
+type EvalContextTestingKnobs struct {
+	// AssertFuncExprReturnTypes indicates whether FuncExpr evaluations
+	// should assert that the returned Datum matches the expected
+	// ReturnType of the function.
+	AssertFuncExprReturnTypes bool
+	// AssertUnaryExprReturnTypes indicates whether UnaryExpr evaluations
+	// should assert that the returned Datum matches the expected
+	// ReturnType of the function.
+	AssertUnaryExprReturnTypes bool
+	// AssertBinaryExprReturnTypes indicates whether BinaryExpr
+	// evaluations should assert that the returned Datum matches the
+	// expected ReturnType of the function.
+	AssertBinaryExprReturnTypes bool
+}
 
 var _ base.ModuleTestingKnobs = &EvalContextTestingKnobs{}
 
@@ -2223,7 +2236,16 @@ func (expr *BinaryExpr) Eval(ctx *EvalContext) (Datum, error) {
 	if right == DNull && !expr.fn.nullableArgs {
 		return DNull, nil
 	}
-	return expr.fn.fn(ctx, left, right)
+	res, err := expr.fn.fn(ctx, left, right)
+	if err != nil {
+		return nil, err
+	}
+	if ctx.TestingKnobs.AssertBinaryExprReturnTypes {
+		if err := ensureExpectedType(expr.fn.ReturnType, res); err != nil {
+			return nil, errors.Wrapf(err, "binary op %q", expr.String())
+		}
+	}
+	return res, err
 }
 
 // Eval implements the TypedExpr interface.
@@ -2965,7 +2987,23 @@ func (expr *FuncExpr) Eval(ctx *EvalContext) (Datum, error) {
 		}
 		return nil, errors.Wrapf(err, "%s()", fName)
 	}
+	if ctx.TestingKnobs.AssertFuncExprReturnTypes {
+		if err := ensureExpectedType(expr.fn.FixedReturnType(), res); err != nil {
+			return nil, errors.Wrapf(err, "function %q", expr.String())
+		}
+	}
 	return res, nil
+}
+
+// ensureExpectedType will return an error if a datum does not match the
+// provided type. If the expected type is Any or if the datum is a Null
+// type, then no error will be returned.
+func ensureExpectedType(exp types.T, d Datum) error {
+	if !(exp.FamilyEqual(types.Any) || d.ResolvedType().Equivalent(types.Null) ||
+		d.ResolvedType().Equivalent(exp)) {
+		return errors.Errorf("expected return type %q, got: %q", exp, d.ResolvedType())
+	}
+	return nil
 }
 
 // Eval implements the TypedExpr interface.
@@ -3083,7 +3121,16 @@ func (expr *UnaryExpr) Eval(ctx *EvalContext) (Datum, error) {
 	if d == DNull {
 		return DNull, nil
 	}
-	return expr.fn.fn(ctx, d)
+	res, err := expr.fn.fn(ctx, d)
+	if err != nil {
+		return nil, err
+	}
+	if ctx.TestingKnobs.AssertUnaryExprReturnTypes {
+		if err := ensureExpectedType(expr.fn.ReturnType, res); err != nil {
+			return nil, errors.Wrapf(err, "unary op %q", expr.String())
+		}
+	}
+	return res, err
 }
 
 // Eval implements the TypedExpr interface.

--- a/pkg/sql/sem/tree/eval.go
+++ b/pkg/sql/sem/tree/eval.go
@@ -1289,7 +1289,7 @@ var BinOps = map[BinaryOperator]binOpOverload{
 		BinOp{
 			LeftType:   types.JSON,
 			RightType:  types.TArray{Typ: types.String},
-			ReturnType: types.JSON,
+			ReturnType: types.String,
 			fn: func(_ *EvalContext, left Datum, right Datum) (Datum, error) {
 				res, err := getJSONPath(*left.(*DJSON), *MustBeDArray(right))
 				if err != nil {

